sdk/digitaltwins/Azure.DigitalTwins.Core/src/Queries/QueryBuilders/Linq/DigitalTwinsQueryBuilder.cs

```diff
@@ -0,0 +1,391 @@
+            string query = string.Empty;
+            Expression e = Evaluator.PartialEval(filter, CanEvaluate);
+            e = ExpressionNormalizer.Normalize(e, new Dictionary<Expression, Expression>());
+
+            LambdaExpression l = e as LambdaExpression;
+            Ensure(l != null);
+            Ensure(l.Parameters.Count == 1);
+
+            query = QueryFilterVisitor.Translate(l.Body);
+
+            return Where(query);
+
+            bool CanEvaluate(Expression e)
+            {
+                if (e is MethodCallExpression call && call.Method.DeclaringType == typeof(DigitalTwinsFunctions))
+                {
+                    return false;
+                }
+
+                return Evaluator.CanBeEvaluatedLocally(e);
+            }
+        }
+
+        private static void Ensure(bool condition, string message = null)
+        {
+            if (!condition)
+            {
+                throw new InvalidOperationException(message ?? "Invalid expression.");
+            }
+        }
+
+        private static string GetPropertyName(Expression<Func<T, object>> selector)
+        {
+            LambdaExpression lambda = selector as LambdaExpression;
+            Ensure(lambda != null);
+            Ensure(lambda.Parameters.Count == 1);
+
+            ParameterExpression param = lambda.Parameters[0];
+            Ensure(param.Type == typeof(T));
+
+            Expression body = lambda.Body;
+            UnaryExpression conversion = body as UnaryExpression;
+
+            if (conversion != null)
+            {
+                Ensure(conversion.NodeType == ExpressionType.Convert);
+                body = conversion.Operand;
+            }
+
+            MemberExpression member = body as MemberExpression;
+            Ensure(member != null);
+            Ensure(member.Expression == param);
+            Ensure(member.Member.MemberType == System.Reflection.MemberTypes.Property);
+
+            return member.Member.Name;
+        }
+
+        /// <summary>
+        /// Gets the string representation of the built query.
+        /// </summary>
+        /// <returns>String represenation of query.</returns>
+        public string GetQueryText()
+        {
+            if (string.IsNullOrEmpty(_queryText))
+            {
+                Build();
+            }
+
+            return _queryText;
+        }
+
+        /// <summary>
+        /// Gets the string representation of the built query.
+        /// </summary>
+        /// <returns>String represenation of query.</returns>
+        public override string ToString() => GetQueryText();
+
+        /// <summary>
+        /// Updates the string representation of the query.
+        /// </summary>
+        /// <returns>Query with updated string representation.</returns>
+        public DigitalTwinsQueryBuilder<T> Build()
+        {
+            QueryAssembler query = new QueryAssembler();
+            SelectClauseAssembler selectClause = _count
+                ? query.SelectCount()
```

Comment:
I don't find this particularly readable; seems like an overuse of the ternary operator. Can we break it down to full `if` statements?

Reply:
I think we can manage that :)